src/uwtools/rocoto.py

```diff
@@ -25,20 +35,111 @@ def _add_jobname(tree: dict) -> None:
             _add_jobname(subtree)
 
 
-# Public functions
-def write_rocoto_xml(input_yaml: str, input_template: str, rendered_output: str) -> None:
+def _add_tasks(
+    input_yaml: OptionalPath = None,
+) -> None:
     """
-    Main entry point.
+    Load YAML config and add job names to each defined workflow task.
 
     :param input_yaml: Path to YAML input file.
-    :param input_template: Path to input template file.
-    :param rendered_output: Path to write rendered XML file.
     """
     values = YAMLConfig(input_yaml)
-    tasks = values["tasks"]
+    tasks = values["workflow"]["tasks"]
     if isinstance(tasks, dict):
         _add_jobname(tasks)
 
+
+def _rocoto_schema_yaml() -> DefinitePath:
+    """
+    The path to the file containing the schema to validate the XML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "rocoto.jsonschema"
+
+
+def _rocoto_schema_xml() -> DefinitePath:
+    """
+    The path to the file containing the Rocoto workflow document template to render.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "schema_with_metatasks.rng"
+
+
+def _write_rocoto_xml(
+    config_file: OptionalPath,
+    rendered_output: DefinitePath,
+) -> None:
+    """
+    Render the given YAML file to XML using the given template.
+
+    :param config_file: Path to YAML input file.
+    :param rendered_output: Path to write rendered XML file.
+    """
+
+    rocoto_template = _rocoto_schema_xml()
```

Comment:
Errors on `rocoto.jinja2` are:
`lxml.etree.XMLSyntaxError: Start tag expected, '<' not found`
`jinja2.exceptions.UndefinedError: 'entities' is undefined`

Reply:
What's the context in which you're seeing those errors? I mean, what are you doing/running when you see those errors?

The `lxml` error looks something like what I'd expect to see if `lxml` was given the RelaxNG schema and asked to validate `rocoto.jinja`, as that file is not well-formed XML and has Jinja2 syntax where `lxml` would expect XML syntax.

The `jinja2` error looks something like what I'd expect to see if an attempt was made to _render_ `rocoto.jinja2` without supplying the necessary data.

Reply:
The first is from `test_rocoto_xml_is_valid` and the second is from `test__write_rocoto_xml` , `test_realize_rocoto_invalid_xml` , and `test_realize_rocoto_xml`. 
Strictly, that write test should be the same as it is currently if it is given rocoto.jinja2, let me look more>
For the validations, its validating `hello_workflow.xml` against the schema. `hello_workflow.xml` is consistent with `hello_workflow.yaml`, so there's something funny in the validation then in reading the schema?

Reply:
Ah. `schema_with_metatasks.rng` is the schema while `rocoto.jinja2` is the template, so I should use all 3. Not sure why 'write' is not working now with that error though

Reply:
If you find yourself blocked, let me know (and push all your changed to your fork) and I'll see if I can replicate the failures and suggest anything.

Reply:
Thanks - I've pushed the update with schema and template and errors. It looks the same so it should work, but I may have been looking at it too long to see it now

Reply:
`rocoto.jinja2` contains the reference `entities.items()`, meaning that the "values" dict passed in to the `J2Template` constructor needs to have a top-level `entities` key. But, IIRC you added a top-level `workflow` key to the test-fixture file `hello_workflow.yaml`, so you might try changing the expression `YAMLConfig(config_file).data` to `YAMLConfig(config_file).data["workflow"]` in `_write_rocoto_xml()`, so that the same `dict` as before is being passed in.

Reply:
I think there's also an error in the updated tests, because when I run `make unittest`, the file `src/uwtools/tests/fixtures/rocoto_invalid.xml` is _changed by the tests_. Test-fixture files should not be modified by tests.

Reply:
whew, got that fixed, not too bad! Now I'm having issues with patching through the XML file, I've updated with my changes

Reply:
Yikes, had to learn a lot about mocking and context managers! However, it works now!